feat: add group/global sorting and search with UI enhancements

MarshuMax opened the pull request:
This PR implements group and global sorting and searching features with corresponding UI improvements. This addresses the feature request raised in issue #1936.

### Features
- Introduced global and group-level sorting with a status bar
- Implemented global search functionality with improved logic
- Added certificate expiration label and persistence of sorting state
- Enhanced UI layout: background styling, component positioning
- Performed language adaptation and canonical annotation cleanups

### Additional Notes
- A sorting option (ascending/descending) and search bar are now available both globally and within individual groups
- Global sorting can be applied while also setting specific sorting rules for one or more selected groups

![image](https://github.com/user-attachments/assets/8f5fff90-0236-400a-ba96-20bda7f3cf60)

Closes #1936

CommanderStorm commented:
Sorry, a +1.1k rewrite of one component implementing different features is not reviewable.
You will need to split this apart.

Looking at the screenshot:
- From the UI side, this does not really fit into our existing style. Look at how buttons are rounded for exampe.
- In terms of functionality, I think this is a bit too complex. This adds multiple search  bars and multiple sorting options.

MarshuMax commented:
Hi @CommanderStorm ,
> Sorry, a +1.1k rewrite of one component implementing different features is not reviewable. You will need to split this apart.

ok, 

> Looking at the screenshot:
> 
> * From the UI side, this does not really fit into our existing style. Look at how buttons are rounded for exampe.

will update

> * In terms of functionality, I think this is a bit too complex. This adds multiple search  bars and multiple sorting options.

do you think we should add search

MarshuMax commented:
Hi @CommanderStorm ,

Thanks for the feedback! 
> Sorry, a +1.1k rewrite of one component implementing different features is not reviewable. You will need to split this apart.

I'll split the PR into smaller, more focused changes. I'll also add some necessary comments to make the logic clearer.

> Looking at the screenshot:
> 
> * From the UI side, this does not really fit into our existing style. Look at how buttons are rounded for exampe.

I'll update the UI to better match the existing style—thanks for pointing that out.

> * In terms of functionality, I think this is a bit too complex. This adds multiple search  bars and multiple sorting options.

Do you have any suggestions regarding the sorting feature?
Also, would it still make sense to include a search function here, or do you think it's unnecessary?

I’ll revise the implementation based on your suggestions and follow up with updated PRs soon. Thanks again!

BR,
Marshu

CommanderStorm commented:
I'd like to simplify this to one button on the top right with just sorting.
No search bar required

We could add it here (nicely out of the way) via a dropdown-button type interaction
![image](https://github.com/user-attachments/assets/c1355823-ccdf-4893-83c9-65e86ec7d03f)

Also note that the state for this should be an url query parameter
=> make it possible to share the exact sorting with others.

MarshuMax commented:
> I'd like to simplify this to one button on the top right (at the group level) with just sorting. No search bar required
> 
> We could add it here (nicely out of the way) via a dropdown-button type interaction ![image](https://private-user-images.githubusercontent.com/26258709/433911406-c1355823-ccdf-4893-83c9-65e86ec7d03f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ3OTE4OTQsIm5iZiI6MTc0NDc5MTU5NCwicGF0aCI6Ii8yNjI1ODcwOS80MzM5MTE0MDYtYzEzNTU4MjMtY2NkZi00ODkzLTgzYzktNjVlODZlYzdkMDNmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE2VDA4MTk1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQzYTgwOGNhMzU1NzFhNGJiYmI4ZDYwNmE5YjJiM2JmMDdlNTc4ZDQ5ZDk4NDFlYWZhZDdkNWRlYTY3ZTI5NGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.acFkXuts1RGmk5ARFIiyupUY655gm1CxLd9qqq75Mck)

I will remove the search functionality, and I believe what you're describing would work like this:
![image](https://github.com/user-attachments/assets/ffc29cb2-fd07-44d3-a7a0-fca6acee65d3)

> 
> Also note that the state for this should be an url query parameter => make it possible to share the exact sorting with others.

I think I generally understand your idea, but I’d like to clarify a few points:
1. Based on your description, it seems the sorting is group-level rather than global.  
Should we remove the global sorting then?

2. Also, is it currently possible to share links with group-level sorting defined?  
If we want to show different sorting for each group, would the URL look something like this:  
`group1_sort=name&group1_order=asc&group2_sort=uptime&group2_order=desc`?

CommanderStorm commented:
> Based on your description, it seems the sorting is group-level rather than global.
Should we remove the global sorting then?

Yes, please do that. That has the addtional advantage of simplifies the PR ^^

> Also, is it currently possible to share links with group-level sorting defined?
If we want to show different sorting for each group, would the URL look something like this:
`group1_sort=name&group1_order=asc`?

That sounds reasonable. Maybe `sort_group1=name_asc` since these are trigged by the same thing?
If simpler to implement the other way around, lets use your version, otherwise my.
That is hard [bikeshedding](https://en.wikipedia.org/wiki/Law_of_triviality) theretorium.

CommanderStorm commented:
> you're describing would work like this

Yes, that is approximately what I was thinking about.
Maybe with an icon instead of text (more out of the way), but that is for you to decide what you like better.

MarshuMax commented:
Hi @CommanderStorm ,

Thank you for your suggestions. Based on your advice, I have made the following adjustments.

## For the adjustments to the sorting
Removed the search functionality and global sorting. The group sorting options are now displayed through a dropdown button, as shown in the image below.
![image](https://github.com/user-attachments/assets/4bd58129-88aa-4686-83d8-1376291b9d65)

I replaced the "Sort by" text on the button with an upward arrow and a downward arrow. When no sorting option is selected, both arrows are gray. When the sorting is in ascending order, the upward arrow turns green; otherwise, the downward arrow turns green. Allow switching between ascending and descending order by clicking the already selected sorting attribute again.

## Support for link-sharing of personal sorting
Implemented the link-sharing feature. 
For example, I have two groups named group1 and group2. I defined the sorting rule for group1 as uptime in ascending order, and group2 as status in descending order. The link would be: 
http://localhost:3000/status/guess?sort_group1=uptime_asc&sort_group2=status_desc. 

When others visit this link, they will see the same sorting effects.
![image](https://github.com/user-attachments/assets/eb3e5a4a-faa4-4012-bb6b-668f131bbba8)

Please check it when you have time. Thanks!

CommanderStorm commented on src/components/PublicGroupList.vue:
```diff
@@ -117,18 +195,263 @@ export default {
+         * @param {object} group object
+         * @returns {object|null} saved sorting settings
+         */
+        getSavedSortSettings(group) {
```
I am a bit torn on this.
I think `localStorage` for this feature is not what I would like - I would like all state to be stored in the query exclusively.
Moving this to the query is also simpler from a code perspective..

Looking at the code below, you have a strong preference.
Is this something which you feel strongly about or would you be fine with using the query state instead?

MarshuMax replied:
You're right, URL state is much cleaner. Removed all localStorage code and now everything goes through `$route.query`. The `sortSettingsFromURL` computed property handles parsing `sort_groupName_asc `format from the URL.

CommanderStorm commented on src/components/PublicGroupList.vue:
```diff
@@ -117,18 +195,263 @@ export default {
+        getSortKey(group) {
+            return group.sortKey || "status";
+        },
+
+        /**
+         * Get sort direction symbol
+         * @param {object} group object
+         * @returns {string} sort direction symbol
+         */
+        getSortDirectionSymbol(group) {
+            return (group.sortDirection === "asc") ? "↑" : "↓";
+        },
```
nit: not used in the code

```suggestion
```

CommanderStorm commented on src/components/PublicGroupList.vue:
```diff
@@ -193,6 +517,100 @@ export default {
+         * Handle browser back/forward button events
+         * @returns {void}
+         */
+        handlePopState() {
+            this.loadSortSettingsFromURL();
```
lets inline this. Much readable what `popstate` does this way.

MarshuMax replied:
Actually went further than inlining - removed the `popstate `listener entirely.
Now using Vue's reactive `sortSettingsFromURL `watcher which automatically responds to route changes. Much cleaner than manual event handling.

CommanderStorm commented on src/components/PublicGroupList.vue:
```diff
@@ -193,6 +517,100 @@ export default {
+                return;
+            }
+
+            const urlParams = new URLSearchParams(window.location.search);
```
I think the vue router already gives access to this.

Handling this manually seems like a lot of unnessesary work, especially in the reactivity department above.

I think the code would be much simpler if instead of hooking up events in the right places, we were to use a `computed` to sort and let vue handle the reactivity.

CommanderStorm commented on src/components/PublicGroupList.vue:
```diff
@@ -10,9 +10,88 @@
+                        <Editable v-model="group.element.name" :contenteditable="editMode" tag="span" data-testid="group-name" />
+                    </div>
+
+                    <div v-if="group.element && group.element.monitorList && group.element.monitorList.length > 1" class="sort-dropdown">
```
thinking about it, could you (if not, I am fine with this) move the code related to the `PublicGroupListSortDropdown` (horrible name, feel free to choose a different one) to its own component?

CommanderStorm reviewed (changes requested):
Looks quite nicely.
I think this needs another round of polishing though.

The main bug is that the query part does not reflect the state..
![image](https://github.com/user-attachments/assets/cce86372-097d-4bd4-96ff-983676bc7f5b)

I have commented in a few other places below too, but those should be minor

MarshuMax commented:
Thanks for the review and detailed feedback! Just saw this I’ll go through the comments and update accordingly soon.

MarshuMax commented on src/components/PublicGroupList.vue:
```diff
@@ -10,9 +10,88 @@
+                        <Editable v-model="group.element.name" :contenteditable="editMode" tag="span" data-testid="group-name" />
+                    </div>
+
+                    <div v-if="group.element && group.element.monitorList && group.element.monitorList.length > 1" class="sort-dropdown">
```
Thanks for noticing!

MarshuMax commented:
Hi @CommanderStorm, do you have any review comments or suggestions for the next steps? Thanks!

CommanderStorm commented:
sorry, I have been super buisy with non-uptime-kuma things.
I won't be able to give a review given the stress that I am currently under from other things.

I would appreciated if other people could test-drive this PR.

CommanderStorm commented:
Sorry for dragging the review out for so long and thanks for the ping.

So the main bug that I would like to see adressed is the bug I noted in the previous review:

<img width="1654" height="954" alt="image" src="https://github.com/user-attachments/assets/2d29567d-76e7-4716-afc6-1938c8de436a" />

That should be resonably simple to fix by either using an already existing ID for this or by passing the ID through.

CommanderStorm commented:
The feature works really nicely and I would love to have it

MarshuMax commented:
> Sorry for dragging the review out for so long and thanks for the ping.
> 
> So the main bug that I would like to see adressed is the bug I noted in the previous review:
> 
> <img alt="image" width="1654" height="954" src="https://private-user-images.githubusercontent.com/26258709/517661842-2d29567d-76e7-4716-afc6-1938c8de436a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjM4NjczNjUsIm5iZiI6MTc2Mzg2NzA2NSwicGF0aCI6Ii8yNjI1ODcwOS81MTc2NjE4NDItMmQyOTU2N2QtNzZlNy00NzE2LWFmYzYtMTkzOGM4ZGU0MzZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTIzVDAzMDQyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMzZmYwNDU4NzUzODIxYzg2ODNhMmI1MDVmMjZhZDJjZGM0M2U2OTlhNTdlMTBhODQ5NDlhNTZkZmMxZjA4NTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.BcITWoFEQblr3QMu8ZWe5bHT9hLUOD_r56Bm1dohUxk">
> That should be resonably simple to fix by either using an already existing ID for this or by passing the ID through.

Thanks for the comment! I've fixed this issue.

The current fix uses the group’s backend ID (or the index if no ID is available) as the unique key, which resolves the issue on the status page where groups with the same name were incorrectly coupled in sorting.
The change is in [src/components/GroupSortDropdown.vue](https://github.com/louislam/uptime-kuma/pull/5766/commits/d13e5316bcf1efca33a2b4508ef6768614d54034)

This is the current display:
<img width="2061" height="1171" alt="image" src="https://github.com/user-attachments/assets/3b791c0f-0dbb-487a-87a9-c18ece7b9616" />

Please bother to help review it again if you have time, and let me know if there are any comments, Thanks!

CommanderStorm reviewed (approved):
Nice work!

Thank you for impelementing this improvement 🎉
